tasks/deploy.py

```diff
@@ -0,0 +1,445 @@
+
+    Args:
+        job_dir (string): path to the job directory
+        build_target (string): eessi-VERSION-COMPONENT-OS-ARCH
+        timestamp (int): timestamp of the tarball
+        repo_name (string): repository of the pull request
+        pr_number (string): number of the pull request
+    """
+    funcname = sys._getframe().f_code.co_name
+
+    tarball = f"{build_target}-{timestamp}.tar.gz"
+    abs_path = os.path.join(job_dir, tarball)
+    log(f"{funcname}(): deploying build '{abs_path}'")
+
+    # obtain config settings
+    deploycfg = config.get_section(DEPLOYCFG)
+    upload_to_s3_script = deploycfg.get(UPLOAD_TO_S3_SCRIPT)
+    endpoint_url = deploycfg.get(ENDPOINT_URL) or ''
+    bucket_name = deploycfg.get(BUCKET_NAME)
+
+    # run 'eessi-upload-to-staging {abs_path}'
+    # (1) construct command line
+    #   script assumes a few defaults:
+    #     bucket_name = 'eessi-staging'
+    #     if endpoint_url not set use EESSI S3 bucket
+    # (2) run command
+    cmd_args = [upload_to_s3_script, ]
+    if len(bucket_name) > 0:
+        cmd_args.extend(['--bucket-name', bucket_name])
+    if len(endpoint_url) > 0:
+        cmd_args.extend(['--endpoint-url', endpoint_url])
+    cmd_args.extend(['--repository', repo_name])
+    cmd_args.extend(['--pull-request', str(pr_number)])
+    cmd_args.append(abs_path)
+    upload_cmd = ' '.join(cmd_args)
+
+    # run_cmd does all the logging we might need
+    out, err, ec = run_cmd(upload_cmd, 'Upload tarball to S3 bucket')
+
+    if ec == 0:
+        # add file to 'job_dir/../uploaded.txt'
+        append_tarball_to_upload_log(tarball, job_dir)
+        # update pr comment
+        update_pr_comment(tarball, repo_name, pr_number, "uploaded",
+                          "succeeded")
+    else:
+        # update pr comment
+        update_pr_comment(tarball, repo_name, pr_number, "not uploaded",
+                          "failed")
+
+
+def uploaded_before(build_target, job_dir):
+    """Determines if a tarball for a job has been uploaded before.
+       Function scans the log named 'job_dir/../uploaded.txt' for
+       the string '.*build_target-.*.tar.gz'.
+
+    Args:
+        build_target (string): eessi-VERSION-COMPONENT-OS-ARCH
+        job_dir (string): directory of the job
+
+    Returns:
+        (string): name of the first tarball found if any or None.
+    """
+    funcname = sys._getframe().f_code.co_name
+
+    log(f"{funcname}(): any previous uploads for {build_target}?")
+
+    pr_base_dir = os.path.dirname(job_dir)
+    uploaded_txt = os.path.join(pr_base_dir, "uploaded.txt")
+
+    if os.path.exists(uploaded_txt):
+        log(f"{funcname}(): upload log '{uploaded_txt}' exists")
+
+        re_string = f".*{build_target}-.*.tar.gz.*"
+        re_build_target = re.compile(re_string)
+
+        with open(uploaded_txt, "r") as uploaded_log:
+            log(f"{funcname}(): scan log for pattern '{re_string}'")
+            for line in uploaded_log:
```

Comment:
Does this actually work?

I would expect that you need to do `uploaded_log.readlines()` here...

Reply:
Seems so. Tested with Python 2.7 and 3.9.